edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs

```diff
@@ -74,12 +101,73 @@ await serviceIdentity
                     .Map(s => this.HandleNewServiceIdentity(s))
                     .GetOrElse(() => this.HandleNoServiceIdentity(id));
             }
+            catch (DeviceInvalidStateException)
+            {
+                await this.HandleNoServiceIdentity(id);
+            }
+            catch (Exception e)
+            {
+                Events.ErrorRefreshingCache(e, id);
+            }
+        }
+
+        bool ShouldNotRefresh(StoredServiceIdentity storedServiceIdentity, bool refreshIfOutOfDate) => !refreshIfOutOfDate || storedServiceIdentity.Timestamp + this.refreshDelay > DateTime.UtcNow;
+
+        void VerifyServiceIdentity(StoredServiceIdentity storedServiceIdentity) => this.VerifyServiceIdentity(storedServiceIdentity.ServiceIdentity);
+
+        void VerifyServiceIdentity(Option<ServiceIdentity> serviceIdentity) => serviceIdentity.ForEach(
+                si =>
+                {
+                    if (si.Status != ServiceIdentityStatus.Enabled)
+                    {
+                        throw new DeviceInvalidStateException("Device is disabled.");
+                    }
+                },
+                () => throw new DeviceInvalidStateException("Device is out of scope."));
+
+        async Task RefreshServiceIdentityInternal(string id)
+        {
+            try
+            {
+                // start refresh
+                Events.RefreshingServiceIdentity(id);
+                // Successfully get response from server
+                Option<ServiceIdentity> serviceIdentity = await this.GetServiceIdentityFromService(id);
+                // If found device, update it otherwise something is wrong with the device, set it as invalid
+                await serviceIdentity
+                    .Map(s => this.HandleNewServiceIdentity(s))
+                    .GetOrElse(() => this.HandleNoServiceIdentity(id));
+                this.VerifyServiceIdentity(serviceIdentity);
+            }
+            catch (DeviceInvalidStateException ex)
+            {
+                Events.ErrorRefreshingCache(ex, id);
+                // Device either out of scope or remove, set it as invalid
+                await this.HandleNoServiceIdentity(id);
+                throw;
+            }
             catch (Exception e)
             {
+                // Refresh failed
                 Events.ErrorRefreshingCache(e, id);
+                throw;
             }
         }
 
+        public async Task VerifyServiceIdentityState(string id, bool refreshIfOutOfDate = false)
+        {
+            Option<StoredServiceIdentity> storedServiceIdentity = await this.GetStoredServiceIdentity(id);
+            // if stored service identity is up to date, use it, otherwise refresh
+            await storedServiceIdentity.Filter(ssi => this.ShouldNotRefresh(ssi, refreshIfOutOfDate))
```

Comment:
We should log here if the cached identity is valid, or if we refreshed it, or if it was just refreshed.. debug logs